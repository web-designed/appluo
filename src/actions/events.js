import uuid from 'uuid'
import moment from 'moment'

//*******************************************************
// EVENTS ACTION GENERATORS
//*******************************************************

const addEvent = (
   {
      note = '', 
      cleaner = 'unknown', 
      cleanedAt = moment().valueOf(),
      createdAt = moment().valueOf(),
      place = '',
      comments = [],
      id = uuid()
   } = {}) => ({
      type:'ADD_EVENT',
      event: {
         id,
         note,
         cleaner,
         place,
         createdAt,
         cleanedAt,
         comments
      }
   }
)

const removeEvent = (id) => ({
   type: 'REMOVE_EVENT',
   id
})

const editEvent = (id, update) => ({
   type: 'EDIT_EVENT', 
   id,
   update
})

const addComment = (eventId, { commentBody, commenter }) => ({
   type: 'ADD_COMMENT',
   eventId,
   comment: {
      id: uuid(),
      createdAt: moment().valueOf(),
      commenter,
      commentBody
   }
})

const removeComment = (eventId, commentId) => ({
   type:'REMOVE_COMMENT',
   eventId,
   commentId
})

const editComment = (eventId, commentId, update) => ({
   type:'EDIT_COMMENT',
   eventId,
   commentId,
   update
})

export { addEvent, removeEvent, editEvent, addComment, removeComment, editComment }