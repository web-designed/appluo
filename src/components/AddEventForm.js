import React from 'react'
import { connect } from 'react-redux'
import uuid from 'uuid'
import moment from 'moment'
import 'react-dates/initialize'
import { SingleDatePicker } from 'react-dates'
import { Link } from 'react-router-dom'

export class AddEventForm extends React.Component {

   state = {
      date: this.props.cleanedAt ? moment(this.props.cleanedAt) : moment(),
      focused: false,
      note: this.props.note ? this.props.note : '',
      place: this.props.place ? this.props.place : this.props.settings.places[0],
      error: '',
      cleaner: this.props.cleaner ? this.props.cleaner : this.props.settings.cleaners[0],
   }

   onCleanerChange = (e) => {
      const cleaner = e.target.value
      this.setState(() => ({ cleaner }))
   }

   onPlaceChange = (e) => {
      const place = e.target.value
      this.setState(() => ({ place }))
   }

   onNoteChange = (e) => {
      const note = e.target.value
      this.setState(() => ({ note }))
   }

   onDateChange = (date) => {
      this.setState(() => ({ date }))
   }

   onFocusChange = ({focused}) => {
      this.setState(() => ({focused}))
   }

   handleSubmit = (e) => {

      e.preventDefault()
      
      const event = {
         cleanedAt: this.state.date.valueOf(),
         createdAt: moment().valueOf(),
         cleaner: this.state.cleaner,
         place: this.state.place,
         note: this.state.note,
      }

      this.props.handleSubmit(event)
   }

   render(){

      console.log(this.props)

      return(
         <div>
            <form onSubmit={this.handleSubmit}>
               <div class="input-group mb-2">
                  <div class="input-group-prepend">
                     <label class="input-group-text" for="cleaner">Name</label>
                  </div>
                  <select 
                     class="custom-select" 
                     id="cleaner"
                     name="cleaner" 
                     onChange={this.onCleanerChange}
                     value={this.state.cleaner.toLowerCase()}
                  >
                     {
                        this.props.settings.cleaners.map( cleaner => {
                           return <option value={cleaner.toLowerCase()} key={cleaner.toLowerCase()}>{cleaner}</option>
                        })
                     }
                  </select>
               </div>
               <div class="input-group mb-2">
                  <div class="input-group-prepend">
                     <label class="input-group-text" for="place">Place</label>
                  </div>
                  <select 
                     class="custom-select" 
                     id="place"
                     name="place" 
                     onChange={this.onPlaceChange}
                     value={this.state.place.toLowerCase()}
                  >
                     {
                        this.props.settings.places.map( place => {
                           return <option value={place.toLowerCase()} key={place}>{place}</option>
                        })
                     }
                  </select>
               </div>
               <div class="input-group mb-2">
                  <div class="input-group-prepend">
                     <label class="input-group-text" for="date">Date</label>
                  </div>
                  <SingleDatePicker 
                     date={this.state.date}
                     onDateChange={this.onDateChange}
                     focused={this.state.focused}
                     onFocusChange={this.onFocusChange}
                     numberOfMonths={1}
                     isOutsideRange={() => false}
                  />
               </div>
               <div class="form-group">
                  <textarea 
                     class="form-control"
                     value={this.state.note}
                     name="note" 
                     placeholder="Add your note"
                     onChange={this.onNoteChange}
                  ></textarea>
               </div>
               <div class="form-group text-right">
                  <Link class="text-danger mr-2" to={'/'}>Cancel</Link>
                  <button class="btn btn-primary">Submit</button>
               </div>
            </form>
         </div>
      )
   }
}

const mapStateToProps = ({ settings }) => ({
   settings: settings
})

export default connect(mapStateToProps)(AddEventForm)