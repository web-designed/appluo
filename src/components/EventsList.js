import React from 'react'
import EventListItem from './EventListItem'
import Sorting from './Sorting'
import { connect } from 'react-redux'
import getVisibleEvents from '../selectors/getVisibleEvents'
import moment from 'moment'

export const EventsList = ({ events } ) => {

   if(events.length > 0){
      let currentMonth
      return (
         <div>
            <Sorting hideFilters={[]} />
            {
               events.map((event) => {
                  
                  const eventsMonth = moment(event.cleanedAt).format('MM')
                  const isNextMonth = eventsMonth !== currentMonth
                  isNextMonth && (currentMonth = eventsMonth)
   
                  return (
                     <div key={event.id}>
                        {isNextMonth && <h3>{moment(event.cleanedAt).format('YYYY-MM')}</h3>}
                        <EventListItem id={event.id} {...event} />
                     </div>
                  )
               })
            }
         </div>
      )
   } else {
      return (
         <div>
            <h3>No Events provided</h3>
         </div>
      )
   }

   
}

const mapStateToProps = ({ events, filters }) => {
   return {
      events: getVisibleEvents(events, filters)
   }
}

export default connect(mapStateToProps)(EventsList)