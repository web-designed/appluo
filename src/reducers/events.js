//*******************************************************
// EVENTS REDUCERS
//*******************************************************

   const eventsReducerDefaultState = []
   const eventsReducer = (state = eventsReducerDefaultState, action) => {
      switch(action.type){
         case 'ADD_EVENT':
            return state.concat(action.event)
         case 'REMOVE_EVENT':
            return state.filter(({ id }) => id !== action.id)
         case 'EDIT_EVENT':
            return state.map((event) => {
               if(event.id === action.id) {
                  return {
                     ...event,
                     ...action.update
                  }
               } else {
                  return event
               }
            })
         case 'ADD_COMMENT':
            return state.map((event) => {
               if(event.id === action.eventId){
                  return {
                     ...event,
                     comments: [...event.comments, action.comment]
                  }
               } else {
                  return event
               }
            })
         case 'REMOVE_COMMENT':
            return state.map((event) => {
               if(event.id === action.eventId){
                  return {
                     ...event,
                     comments: event.comments.filter((comment) => {
                        return comment.id !== action.commentId
                     })
                  }
               } else {
                  return event
               }
            })
         case 'EDIT_COMMENT':
            return state.map((event) => {
               if(event.id === action.eventId){
                  return {
                     ...event,
                     comments: event.comments.map((comment) => {
                        if(comment.id === action.commentId){
                           return {
                              ...comment,
                              ...action.update
                           }
                        } else {
                           return comment
                        }
                     }) 
                  }
               } else {
                  return event
               }
            })
         default:
            return state
      }
   }

export default eventsReducer