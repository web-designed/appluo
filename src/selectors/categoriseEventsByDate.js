import moment from 'moment'

const categoriseEventsByDate = (events) => {      
   let data = {}
   let currentYear
   let currentMonth

   events.map( event => {
      
      const eventsYear = moment(event.cleanedAt).format('Y')
      const eventsMonth = moment(event.cleanedAt).format('MM')
      const isNextYear = eventsYear !== currentYear
      const isNextMonth = eventsMonth !== currentMonth

      if(isNextYear){
         currentYear = eventsYear
         data[eventsYear] = {}
      }

      if(isNextMonth){
         currentMonth = eventsMonth
         data[eventsYear][eventsMonth] = []
      }

      data[eventsYear][eventsMonth] = [
         ...data[eventsYear][eventsMonth],
         event
      ]
   })
   return data
}

export default categoriseEventsByDate