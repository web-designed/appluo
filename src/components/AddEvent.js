import React from 'react'
import AddEventForm from './AddEventForm'

const AddEvent = () => {
   return(
      <div>
         <h1>Add a cleaning event</h1>
         <AddEventForm />
      </div>
   )
}

export default AddEvent