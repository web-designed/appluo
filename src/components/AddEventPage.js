import React from 'react'
import AddEventForm from './AddEventForm'
import { addEvent } from '../actions/events'
import { connect } from 'react-redux'

const AddEventPage = (props) => (
   <div>
      <h1>Add a cleaning</h1>
      <AddEventForm 
         handleSubmit={ (event) => {
            props.onSubmit(event)
            props.history.push('/')
         }} 
         history={props.history} 
      />
   </div>
)

const mapDispatchToProps = (dispatch) => ({
   onSubmit: (expense) => dispatch(addEvent(expense))
})

export default connect(undefined, mapDispatchToProps)(AddEventPage)