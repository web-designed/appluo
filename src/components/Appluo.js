import React from 'react'
import EventsList from './EventsList'

class Appluo extends React.Component {
   render() {
      return (
         <div>
            <EventsList />
         </div>
      )
   }
}

export default Appluo