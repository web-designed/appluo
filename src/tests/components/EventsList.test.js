import React from 'react'
import { shallow } from 'enzyme'
import { EventsList } from '../../components/EventsList'
import events from '../fixtures/events'
import categoriseEventsByDate from '../../selectors/categoriseEventsByDate'

test('should render EventsList correctly', () => {
   const eventsCategorised = categoriseEventsByDate(events)
   const wrapper = shallow(<EventsList events={categoriseEventsByDate} />)
   expect(wrapper).toMatchSnapshot()
})

test('should render EventsList correctly if no events provided', () => {
   const wrapper = shallow(<EventsList events={[]} />)
   expect(wrapper).toMatchSnapshot()
})