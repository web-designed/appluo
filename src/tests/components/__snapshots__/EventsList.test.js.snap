// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EventsList correctly 1`] = `
<div>
  <Connect(Sorting)
    hideFilters={Array []}
  />
  <div
    key="0"
  >
    <h3>
      1970-01
    </h3>
    <Connect(EventListItem)
      cleanedAt={0}
      cleaner="Kylu"
      comments={
        Array [
          Object {
            "commentBody": "this is my comment",
            "commenter": "Kylu",
            "createdAt": 125,
            "id": "1",
          },
          Object {
            "commentBody": "this is my another comment",
            "commenter": "Chris",
            "createdAt": 130,
            "id": "2",
          },
        ]
      }
      createdAt={0}
      id="0"
      note="Alles sauber gemacht"
      place="Bad"
    />
  </div>
  <div
    key="1"
  >
    <h3>
      1969-12
    </h3>
    <Connect(EventListItem)
      cleanedAt={-345600000}
      cleaner="Alex"
      comments={
        Array [
          Object {
            "commentBody": "this is my comment",
            "commenter": "Kylu",
            "createdAt": 125,
            "id": "1",
          },
          Object {
            "commentBody": "this is my another comment",
            "commenter": "Chris",
            "createdAt": 130,
            "id": "2",
          },
        ]
      }
      createdAt={-345600000}
      id="1"
      note="Alles sauber gemacht"
      place="KÃ¼che"
    />
  </div>
  <div
    key="2"
  >
    <h3>
      1970-01
    </h3>
    <Connect(EventListItem)
      cleanedAt={345600000}
      cleaner="Till"
      comments={
        Array [
          Object {
            "commentBody": "this is my comment",
            "commenter": "Kylu",
            "createdAt": 125,
            "id": "1",
          },
          Object {
            "commentBody": "this is my another comment",
            "commenter": "Chris",
            "createdAt": 130,
            "id": "2",
          },
        ]
      }
      createdAt={345600000}
      id="2"
      note="Alles sauber gemacht"
      place="Flur"
    />
  </div>
</div>
`;

exports[`should render EventsList correctly ifno events provided 1`] = `
<div>
  <h3>
    No Events provided
  </h3>
</div>
`;
